### ISSUE TITLE
crucible-jvm: Investigate bugs when verifying JDK 9+ code that uses String

### ISSUE THREAD

**Issue by RyanGlScott** [GaloisInc/crucible#641]
There are still some yet-to-be-diagnosed issues with using `crucible-jvm` in tandem with JDK 9 or later. I'll copy-paste a relevant excerpt from https://github.com/GaloisInc/saw-script/issues/861#issuecomment-774266407, which describes the challenge:

> * [...] After bumping the submodule to commit [GaloisInc/jvm-parser@d440e6b](https://github.com/GaloisInc/jvm-parser/commit/d440e6bc91c9ba7d48f04e3a7cbf3086d9720b1e), which contains a fix for that issue, I instead get a different error when running the `crucible-jvm` test suite:
>   
>   ```
>   javac -g Str.java
>   Compiled from "Str.java"
>   class Str {
>     Str();
>       Code:
>          0: aload_0
>          1: invokespecial #1                  // Method java/lang/Object."<init>":()V
>          4: return
>   
>     public static void main(java.lang.String[]);
>       Code:
>          0: ldc           #2                  // String 1234567890
>          2: astore_1
>          3: new           #3                  // class java/lang/StringBuffer
>          6: dup
>          7: aload_1
>          8: invokespecial #4                  // Method java/lang/StringBuffer."<init>":(Ljava/lang/String;)V
>         11: astore_2
>         12: aload_2
>         13: iconst_2
>         14: ldc           #5                  // String abcd
>         16: invokevirtual #6                  // Method java/lang/StringBuffer.insert:(ILjava/lang/String;)Ljava/lang/StringBuffer;
>         19: pop
>         20: getstatic     #7                  // Field java/lang/System.out:Ljava/io/PrintStream;
>         23: aload_2
>         24: invokevirtual #8                  // Method java/io/PrintStream.println:(Ljava/lang/Object;)V
>         27: return
>   }
>   [Crux] Checking "Str.java"
>   starting executeCrucibleJVM
>   initializeClass java/lang/String  (start)
>   initializeClass java/lang/Object  (start)
>   special java/lang/Object/<clinit>
>   initializeClass java/lang/Object  (finish)
>   special java/lang/String/<clinit>
>   initializeClass java/lang/String  (finish)
>   [Crux] resolveField: Field FieldId {fieldIdClass = ClassName "java/lang/Object", fieldIdName = "value", fieldIdType = char[]} not found
>   CallStack (from HasCallStack):
>     error, called at src/Lang/Crucible/JVM/Translation/Monad.hs:61:15 in crucible-jvm-0.2-inplace:Lang.Crucible.JVM.Translation.Monad
>     jvmFail, called at src/Lang/Crucible/JVM/Translation/Class.hs:836:23 in crucible-jvm-0.2-inplace:Lang.Crucible.JVM.Translation.Class
>   ```
>   
>   
>   
>   This time, the problem lies in the fact that the API for the `String` class has changed. In JDK 8, the private `value` field in `String` [has type `char[]`](https://github.com/openjdk/jdk/blob/9a9add8825a040565051a09010b29b099c2e7d49/jdk/src/share/classes/java/lang/String.java#L114), but in recent JDKs, such as JDK11, its type has changed to [`byte[]`](https://github.com/openjdk/jdk/blob/da75f3c4ad5bdf25167a3ed80e51f567ab3dbd01/src/java.base/share/classes/java/lang/String.java#L140). `crucible-jvm`, on the other hand, hard-codes the assumption that `value` is of type `char[]` here:
>   https://github.com/GaloisInc/crucible/blob/12a1fba8b31073d8f7badd3718f8de03f7244d14/crucible-jvm/src/Lang/Crucible/JVM/Translation/Class.hs#L907-L915
> 
> * As a quick experiment, I tried to see what would happen if I updated this assumption to use the modern JDK's type for `value`:
>   ```diff
>   diff --git a/crucible-jvm/src/Lang/Crucible/JVM/Translation/Class.hs b/crucible-jvm/src/Lang/Crucible/JVM/Translation/Class.hs
>   index 3c09eb72..821ab6cd 100644
>   --- a/crucible-jvm/src/Lang/Crucible/JVM/Translation/Class.hs
>   +++ b/crucible-jvm/src/Lang/Crucible/JVM/Translation/Class.hs
>   @@ -911,7 +911,7 @@ refFromString s =  do
>      -- We'll want want to REVISIT this in the future.
>      obj1 <- setInstanceFieldValue
>        obj
>   -    (J.FieldId "java/lang/String" "value" J.charArrayTy)
>   +    (J.FieldId "java/lang/String" "value" J.byteArrayTy)
>        (RValue (App (JustValue knownRepr arrRef)))
>    
>      obj2 <- setInstanceFieldValue
>   ```
>   
>   
>   This appeared to fix the previous issue, but it uncovered another issue:
>   
>   ```
>   javac -g Str.java
>   Compiled from "Str.java"
>   class Str {
>     Str();
>       Code:
>          0: aload_0
>          1: invokespecial #1                  // Method java/lang/Object."<init>":()V
>          4: return
>   
>     public static void main(java.lang.String[]);
>       Code:
>          0: ldc           #2                  // String 1234567890
>          2: astore_1
>          3: new           #3                  // class java/lang/StringBuffer
>          6: dup
>          7: aload_1
>          8: invokespecial #4                  // Method java/lang/StringBuffer."<init>":(Ljava/lang/String;)V
>         11: astore_2
>         12: aload_2
>         13: iconst_2
>         14: ldc           #5                  // String abcd
>         16: invokevirtual #6                  // Method java/lang/StringBuffer.insert:(ILjava/lang/String;)Ljava/lang/StringBuffer;
>         19: pop
>         20: getstatic     #7                  // Field java/lang/System.out:Ljava/io/PrintStream;
>         23: aload_2
>         24: invokevirtual #8                  // Method java/io/PrintStream.println:(Ljava/lang/Object;)V
>         27: return
>   }
>   Warning: Parsing the index cache failed (Data.Binary.Get.runGet at position
>   16: Non-matching structured hashes: f46da61e7afa58a5e8fd1d2b6fb79899;
>   expected: a257ca064dfb5e0cb74f74e64a975b9e). Trying to regenerate the index
>   cache...
>   Resolving dependencies...
>   [Crux] Checking "Str.java"
>   starting executeCrucibleJVM
>   initializeClass java/lang/String  (start)
>   initializeClass java/lang/Object  (start)
>   special java/lang/Object/<clinit>
>   initializeClass java/lang/Object  (finish)
>   special java/lang/String/<clinit>
>   initializeClass java/lang/String  (finish)
>   new ClassName "java/lang/StringBuffer" (start)
>   fields are [FieldId {fieldIdClass = ClassName "java/lang/AbstractStringBuilder", fieldIdName = "value", fieldIdType = byte[]},FieldId {fieldIdClass = ClassName "java/lang/AbstractStringBuilder", fieldIdName = "coder", fieldIdType = byte},FieldId {fieldIdClass = ClassName "java/lang/AbstractStringBuilder", fieldIdName = "count", fieldIdType = int},FieldId {fieldIdClass = ClassName "java/lang/AbstractStringBuilder", fieldIdName = "EMPTYVALUE", fieldIdType = byte[]},FieldId {fieldIdClass = ClassName "java/lang/AbstractStringBuilder", fieldIdName = "MAX_ARRAY_SIZE", fieldIdType = int},FieldId {fieldIdClass = ClassName "java/lang/StringBuffer", fieldIdName = "toStringCache", fieldIdType = java.lang.String},FieldId {fieldIdClass = ClassName "java/lang/StringBuffer", fieldIdName = "serialVersionUID", fieldIdType = long},FieldId {fieldIdClass = ClassName "java/lang/StringBuffer", fieldIdName = "serialPersistentFields", fieldIdType = java.io.ObjectStreamField[]}]
>   new ClassName "java/lang/StringBuffer" (finish)
>   initializeClass java/lang/StringBuffer  (start)
>   initializeClass java/lang/AbstractStringBuilder  (start)
>   initializeClass java/lang/Object  (start)
>   initializeClass java/lang/Object  (finish)
>   invoking static method with return type UnitRepr
>   initializeClass java/lang/AbstractStringBuilder  (finish)
>   special java/lang/StringBuffer/<clinit>
>   initializeClass java/lang/StringBuffer  (finish)
>   invoking static method with return type UnitRepr
>   initializeClass java/lang/String  (start)
>   initializeClass java/lang/Object  (start)
>   initializeClass java/lang/Object  (finish)
>   initializeClass java/lang/String  (finish)
>   invoke: java/lang/StringBuffer/insert
>   finish invoke:java/lang/StringBuffer/insert
>   initializeClass java/lang/System  (start)
>   initializeClass java/lang/Object  (start)
>   initializeClass java/lang/Object  (finish)
>   Initializing System.out static field
>   Finished initializing System.out
>   Initializing System.err static field
>   Finished initializing System.err
>   initializeClass java/lang/System  (finish)
>   invoke: java/io/PrintStream/println
>   finish invoke:java/io/PrintStream/println
>   initializeClass java/lang/AbstractStringBuilder  (start)
>   initializeClass java/lang/Object  (start)
>   initializeClass java/lang/Object  (finish)
>   invoking static method with return type UnitRepr
>   initializeClass java/lang/AbstractStringBuilder  (finish)
>   invoke: java/lang/String/length
>   finish invoke:java/lang/String/length
>   initializeClass java/lang/AbstractStringBuilder  (start)
>   initializeClass java/lang/Object  (start)
>   initializeClass java/lang/Object  (finish)
>   invoking static method with return type UnitRepr
>   initializeClass java/lang/AbstractStringBuilder  (finish)
>   invoking static method with return type UnitRepr
>   invoke: java/lang/StringBuffer/append
>   finish invoke:java/lang/StringBuffer/append
>   getfield "java/lang/String.value"
>   invoke: java/lang/String/coder
>   finish invoke:java/lang/String/coder
>   initializeClass java/lang/String  (start)
>   initializeClass java/lang/Object  (start)
>   initializeClass java/lang/Object  (finish)
>   initializeClass java/lang/String  (finish)
>   getfield "java/lang/String.coder"
>   initializeClass java/lang/Object  (start)
>   initializeClass java/lang/Object  (finish)
>   invoking static method with return type UnitRepr
>   initializeClass java/lang/String  (start)
>   initializeClass java/lang/Object  (start)
>   initializeClass java/lang/Object  (finish)
>   initializeClass java/lang/String  (finish)
>   initializeClass java/lang/StringUTF16  (start)
>   initializeClass java/lang/Object  (start)
>   initializeClass java/lang/Object  (finish)
>   invoking static method with return type UnitRepr
>   initializeClass java/lang/StringUTF16  (finish)
>   invoking static method with return type MaybeRepr (ReferenceRepr (RecursiveRepr JVM_object []))
>   putfield "java/lang/AbstractStringBuilder.value"
>   putfield "java/lang/AbstractStringBuilder.coder"
>   putfield "java/lang/AbstractStringBuilder.value"
>   putfield "java/lang/AbstractStringBuilder.coder"
>   invoke: java/lang/Class/desiredAssertionStatus
>   finish invoke:java/lang/Class/desiredAssertionStatus
>   initializeClass java/lang/StringUTF16  (start)
>   initializeClass java/lang/Object  (start)
>   initializeClass java/lang/Object  (finish)
>   invoking static method with return type UnitRepr
>   initializeClass java/lang/StringUTF16  (finish)
>   initializeClass java/lang/StringUTF16  (start)
>   initializeClass java/lang/Object  (start)
>   initializeClass java/lang/Object  (finish)
>   invoking static method with return type UnitRepr
>   initializeClass java/lang/StringUTF16  (finish)
>   invoking static method with return type BVRepr 32
>   initializeClass java/lang/StringUTF16  (start)
>   initializeClass java/lang/Object  (start)
>   initializeClass java/lang/Object  (finish)
>   invoking static method with return type UnitRepr
>   initializeClass java/lang/StringUTF16  (finish)
>   initializeClass java/lang/StringUTF16  (start)
>   initializeClass java/lang/Object  (start)
>   initializeClass java/lang/Object  (finish)
>   invoking static method with return type UnitRepr
>   initializeClass java/lang/StringUTF16  (finish)
>   initializeClass java/lang/StringUTF16  (start)
>   initializeClass java/lang/Object  (start)
>   initializeClass java/lang/Object  (finish)
>   invoking static method with return type UnitRepr
>   initializeClass java/lang/StringUTF16  (finish)
>   initializeClass java/lang/StringUTF16  (start)
>   initializeClass java/lang/Object  (start)
>   initializeClass java/lang/Object  (finish)
>   invoking static method with return type UnitRepr
>   initializeClass java/lang/StringUTF16  (finish)
>   Abort due to false assumption:
>     internal: error: in java/lang/StringUTF16.<clinit>
>     null dereference
>     In java/lang/StringUTF16.<clinit> at internal
>     When calling <clinit>
>     In java/lang/AbstractStringBuilder.<init> at internal
>     When calling <init>
>     In java/lang/StringBuffer.<init> at internal
>     When calling <init>
>     In Str.main at internal
>     When calling main
>   [Crux] Attempting to prove verification conditions.
>   [Crux] Simulation complete.
>   [Crux] Found counterexample for verification goal
>   
>   [Crux] Goal status:
>   [Crux]   Total: 1
>   [Crux]   Proved: 0
>   [Crux]   Disproved: 1
>   [Crux]   Incomplete: 0
>   [Crux]   Unknown: 0
>   [Crux] Overall status: Invalid.
>   ```
>   
>   
>   
>   A null dereference! I couldn't tell at a glance what was causing it, but it's quite possible that this arises due to other hard-coded assumptions in `crucible-jvm` that no longer hold in modern JDKs.

**Comment by RyanGlScott** [comment]
Indeed, this is a pretty notable limitation of `crucible-jvm`'s support for JDK9+. In order to make this work, I think the first step would be to characterize all of the differences in how `String.class` is implemented in JDK9+ versus older JDKs and audit which of the changes affect our model of the class.

**Comment by WeeknightMVP** [comment]
Thanks!  The same issue is encountered in (my installation of) OpenJDK 8 unless specifying `-b $JAVA_BIN`, which might be related...

**Comment by RyanGlScott** [comment]
Hm, I am surprised to hear that this happens even with OpenJDK 8. Just to be sure, are you putting OpenJDK 8 first on your `PATH` before any other JDK versions?

**Comment by WeeknightMVP** [comment]
> Hm, I am surprised to hear that this happens even with OpenJDK 8. Just to be sure, are you putting OpenJDK 8 first on your `PATH` before any other JDK versions?

Aha.  I did not modify `PATH` when trying different versions, so this would have remained OpenJDK 17 throughout:

```
$ which javac
/usr/bin/javac
$ readlink -f $(which javac)
/usr/lib/jvm/java-17-openjdk-amd64/bin/javac
```

**Comment by RyanGlScott** [comment]
Right, the `PATH` is where SAW looks by default when `--java-bin-dirs` is not set. See [this section](https://saw.galois.com/manual.html#invoking-saw) of the SAW manual (under "`PATH`").
